contracts/src/ERC20Destination.sol

```diff
@@ -92,14 +108,55 @@ contract ERC20Destination is IERC20Bridge, TeleporterTokenDestination, ERC20 {
      *
      * Calls {ERC20-_burn} to burn tokens from this contract.
      */
-    function _burn(uint256 amount) internal virtual override {
+    function _burn(uint256 amount) internal override {
         _burn(address(this), amount);
     }
 
     /**
-     * @dev See {TeleporterTokenDestination-_scaleTokens}
+     * @dev See {TeleporterTokenDestination-_handleSendAndCall}
+     *
+     * Mints the tokens to this contract, approves the recipient contract to spend them,
+     * and calls {IERC20SendAndCallReceiver-receiveTokens} on the recipient contract.
+     * If the call fails or doesn't spend all of the tokens, the remaining amount is
+     * sent to the fallback recipient.
      */
-    function _scaleTokens(uint256 value, bool) internal pure override returns (uint256) {
-        return value;
+    function _handleSendAndCall(
+        SingleHopCallMessage memory message,
+        uint256 amount
+    ) internal override {
+        // Mint the tokens to this contract address.
+        _mint(address(this), amount);
+
+        // Approve the destination contract to spend the amount.
+        _approve(address(this), message.recipientContract, amount);
```

Comment:
Why do we use `_approve` here, but `SafeERC20.safeApprove` in `ERC20Source.sol`?

Reply:
Same for the use of `_transfer` vs `SafeERC20.safeTransfer`

Reply:
For `ERC20Source`, each of those methods is an external call to an independent ERC20 contract, but in the case of the `Destination` contracts, they are internal calls (since `ERC20Destination` is itself the ERC20 implementation).

Reply:
Gotcha. And `_approve`/`_transfer` are preferred for internal calls for gas considerations, correct?

Reply:
Yeah that's correct. The `SafeERC20` library calls are also used to handle calls to external contracts that could have various implementations (i.e. some ERC20 implementations revert if there is insufficient balance, others just return value), but since this is an internal call to known code, that is not necessary.

Reply:
It also saves on gas as you said, because otherwise it would be making an external call equivalent to `address(this).approve`.